File: apps/ios/SignalASI/CloudImagePayload.swift
import Foundation

struct CloudImagePayload: Equatable {
  static let maximumBytes = 64 * 1_024 * 1_024

  var displayName: String
  var mimeType: String
  var data: Data

  init(displayName: String, mimeType: String, data: Data) throws {
    let normalizedMimeType = mimeType.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    guard !data.isEmpty,
          data.count <= Self.maximumBytes,
          normalizedMimeType.hasPrefix("image/") else {
      throw CloudImagePayloadError.invalidPayload(displayName)
    }
    self.displayName = SignalASIAttachmentPayloadBuilder.sanitizeName(displayName)
    self.mimeType = normalizedMimeType
    self.data = data
  }

  var base64: String {
    data.base64EncodedString()
  }
}

enum CloudImagePayloadError: LocalizedError, Equatable {
  case invalidPayload(String)
  case preparationFailed(String)

  var errorDescription: String? {
    switch self {
    case .invalidPayload(let name):
      return "Cloud image payload is invalid: \(name)"
    case .preparationFailed(let name):
      return "Cloud image could not be prepared: \(name)"
    }
  }
}

enum CloudImagePayloadFactory {
  static func prepare(_ attachments: [SignalASIDraftAttachment]) throws -> [CloudImagePayload] {
    try attachments
      .filter(\.isImage)
      .map { attachment in
        guard let encoded = AgentMediaAttachmentTransportEncoder.inlinePayload(
          for: attachment,
          profile: nil,
          remainingBytes: CloudImagePayload.maximumBytes
        ) else {
          throw CloudImagePayloadError.preparationFailed(attachment.displayName)
        }
        return try CloudImagePayload(
          displayName: encoded.displayName,
          mimeType: encoded.mimeType,
          data: encoded.data
        )
      }
  }
}

enum CloudVisionPayloadEncoder {
  static func attachOpenAI(
    to conversation: inout [[String: Any]],
    images: [CloudImagePayload]
  ) {
    guard !images.isEmpty else { return }
    let index = latestUserIndex(in: conversation) ?? {
      conversation.append(["role": "user", "content": ""])
      return conversation.count - 1
    }()
    var message = conversation[index]
    var content = contentParts(message["content"], textType: "text")
    for image in images {
      content.append([
        "type": "image_url",
        "image_url": [
          "url": "data:\(image.mimeType);base64,\(image.base64)",
          "detail": "auto"
        ]
      ])
    }
    message["content"] = content
    conversation[index] = message
  }

  static func attachAnthropic(
    to conversation: inout [[String: Any]],
    images: [CloudImagePayload]
  ) {
    guard !images.isEmpty else { return }
    let index = latestUserIndex(in: conversation) ?? {
      conversation.append(["role": "user", "content": ""])
      return conversation.count - 1
    }()
    var message = conversation[index]
    var content = contentParts(message["content"], textType: "text")
    for image in images {
      content.append([
        "type": "image",
        "source": [
          "type": "base64",
          "media_type": image.mimeType,
          "data": image.base64
        ]
      ])
    }
    message["content"] = content
    conversation[index] = message
  }

  static func attachGemini(
    to conversation: inout [[String: Any]],
    images: [CloudImagePayload]
  ) {
    guard !images.isEmpty else { return }
    let index = latestUserIndex(in: conversation) ?? {
      conversation.append(["role": "user", "parts": [[String: Any]]()])
      return conversation.count - 1
    }()
    var message = conversation[index]
    var parts = message["parts"] as? [[String: Any]] ?? []
    for image in images {
      parts.append([
        "inline_data": [
          "mime_type": image.mimeType,
          "data": image.base64
        ]
      ])
    }
    message["parts"] = parts
    conversation[index] = message
  }

  private static func latestUserIndex(in conversation: [[String: Any]]) -> Int? {
    conversation.indices.reversed().first { conversation[$0]["role"] as? String == "user" }
  }

  private static func contentParts(_ value: Any?, textType: String) -> [[String: Any]] {
    if let content = value as? [[String: Any]] {
      return content
    }
    if let content = value as? [[Any]] {
      return content.compactMap { $0 as? [String: Any] }
    }
    guard let text = value as? String, !text.isEmpty else { return [] }
    return [["type": textType, "text": text]]
  }
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/CloudImagePayload.kt
package com.galaxyssi.chat

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.util.Base64

internal data class CloudImagePayload(
    val displayName: String,
    val mimeType: String,
    val bytes: ByteArray
) {
    init {
        require(bytes.isNotEmpty()) { "Cloud image payload must not be empty" }
        require(bytes.size <= MAX_BYTES) { "Cloud image payload exceeds the 100 KB limit" }
        require(mimeType.startsWith("image/")) { "Cloud image payload must use an image MIME type" }
    }

    fun base64(): String = Base64.getEncoder().encodeToString(bytes)

    companion object {
        const val MAX_BYTES = AgentImagePipeline.TARGET_TRANSPORT_BYTES
    }
}

internal object CloudImagePayloadFactory {
    fun prepare(
        context: Context,
        attachments: List<AgentInputAttachment>
    ): List<CloudImagePayload> = attachments
        .filter(AgentInputAttachment::isImage)
        .map { attachment ->
            val encoded = AgentImagePipeline.encodeForTransport(
                context = context,
                attachment = attachment,
                byteLimit = CloudImagePayload.MAX_BYTES
            ) ?: error(context.getString(R.string.cloud_image_prepare_failed, attachment.displayName))
            CloudImagePayload(
                displayName = encoded.transportName(attachment.displayName),
                mimeType = encoded.mimeType,
                bytes = encoded.bytes
            )
        }
}

/** Adds compressed image content only to the latest user turn. */
internal object CloudVisionPayloadEncoder {
    fun attachOpenAi(conversation: JSONArray, images: List<CloudImagePayload>) {
        if (images.isEmpty()) return
        val message = latestRole(conversation, "user") ?: JSONObject()
            .put("role", "user")
            .also(conversation::put)
        val content = contentArray(message, "text")
        images.forEach { image ->
            content.put(
                JSONObject()
                    .put("type", "image_url")
                    .put(
                        "image_url",
                        JSONObject()
                            .put("url", "data:${image.mimeType};base64,${image.base64()}")
                            .put("detail", "auto")
                    )
            )
        }
        message.put("content", content)
    }

    fun attachAnthropic(conversation: JSONArray, images: List<CloudImagePayload>) {
        if (images.isEmpty()) return
        val message = latestRole(conversation, "user") ?: JSONObject()
            .put("role", "user")
            .also(conversation::put)
        val content = contentArray(message, "text")
        images.forEach { image ->
            content.put(
                JSONObject()
                    .put("type", "image")
                    .put(
                        "source",
                        JSONObject()
                            .put("type", "base64")
                            .put("media_type", image.mimeType)
                            .put("data", image.base64())
                    )
            )
        }
        message.put("content", content)
    }

    fun attachGemini(conversation: JSONArray, images: List<CloudImagePayload>) {
        if (images.isEmpty()) return
        val message = latestRole(conversation, "user") ?: JSONObject()
            .put("role", "user")
            .put("parts", JSONArray())
            .also(conversation::put)
        val parts = message.optJSONArray("parts") ?: JSONArray().also { replacement ->
            message.optString("parts").takeIf(String::isNotBlank)?.let { text ->
                replacement.put(JSONObject().put("text", text))
            }
            message.put("parts", replacement)
        }
        images.forEach { image ->
            parts.put(
                JSONObject().put(
                    "inline_data",
                    JSONObject()
                        .put("mime_type", image.mimeType)
                        .put("data", image.base64())
                )
            )
        }
    }

    private fun latestRole(conversation: JSONArray, role: String): JSONObject? {
        for (index in conversation.length() - 1 downTo 0) {
            val message = conversation.optJSONObject(index) ?: continue
            if (message.optString("role") == role) return message
        }
        return null
    }

    private fun contentArray(message: JSONObject, textType: String): JSONArray {
        val existing = message.opt("content")
        if (existing is JSONArray) return existing
        return JSONArray().apply {
            val text = when (existing) {
                null, JSONObject.NULL -> ""
                else -> existing.toString()
            }
            if (text.isNotBlank()) {
                put(JSONObject().put("type", textType).put("text", text))
            }
        }
    }
}
